File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/status/StatusBarModel.kt
package com.tyhp.lang.status

import com.tyhp.lang.binary.BinaryStatus
import com.tyhp.lang.debug.ProxyRunState
import com.tyhp.lang.lsp.LspClientState

data class StatusBarView(
    val text: String,
    val tooltip: String,
    val error: Boolean,
    val warning: Boolean,
)

data class StatusBarInput(
    val projectLabel: String,
    val lspState: LspClientState,
    val binaryStatus: BinaryStatus,
    val binaryPath: String? = null,
    val binaryMessage: String? = null,
    val proxyState: ProxyRunState = ProxyRunState.STOPPED,
    val proxyDetail: String? = null,
)

/**
 * Compact status-bar copy: `Tyhp` + project / “no project” + LSP + binary health
 * + XDebug proxy listening state.
 */
fun formatStatusBar(input: StatusBarInput): StatusBarView {
    val project = input.projectLabel.trim().ifEmpty { "not in a Tyhp project" }
    val lsp = lspLabel(input.lspState)
    val binaryOk = input.binaryStatus == BinaryStatus.OK
    val proxy = proxyLabel(input.proxyState, input.proxyDetail)
    val tooltipLines = listOf(
        "Project: $project",
        "Language server: $lsp",
        "CLI: ${if (binaryOk) input.binaryPath ?: "ok" else input.binaryMessage ?: "missing"}",
        "XDebug proxy: $proxy",
        "Click for Tyhp actions.",
    )
    val tooltip = tooltipLines.joinToString("\n")

    if (!binaryOk) {
        return StatusBarView(
            text = "Tyhp · $project · CLI missing",
            tooltip = tooltip,
            error = true,
            warning = false,
        )
    }

    if (input.lspState == LspClientState.ERROR) {
        return StatusBarView(
            text = "Tyhp · $project · LSP error",
            tooltip = tooltip,
            error = true,
            warning = false,
        )
    }

    if (input.lspState == LspClientState.STARTING) {
        return StatusBarView(
            text = "Tyhp · $project · starting",
            tooltip = tooltip,
            error = false,
            warning = false,
        )
    }

    if (input.lspState == LspClientState.STOPPED) {
        return StatusBarView(
            text = "Tyhp · $project · LSP stopped",
            tooltip = tooltip,
            error = false,
            warning = isMissingProject(project),
        )
    }

    if (input.proxyState == ProxyRunState.ERROR) {
        return StatusBarView(
            text = "Tyhp · $project · proxy error",
            tooltip = tooltip,
            error = false,
            warning = true,
        )
    }

    return StatusBarView(
        text = "Tyhp · $project · $lsp${proxyTextSuffix(input.proxyState)}",
        tooltip = tooltip,
        error = false,
        warning = isMissingProject(project),
    )
}

/** Status-bar popup actions for the current proxy lifecycle state. */
fun proxyStatusActions(state: ProxyRunState): List<String> {
    val actions = mutableListOf<String>()
    if (state != ProxyRunState.RUNNING && state != ProxyRunState.STARTING) {
        actions.add("start")
    }
    if (state == ProxyRunState.RUNNING || state == ProxyRunState.STARTING || state == ProxyRunState.STOPPING) {
        actions.add("stop")
        actions.add("restart")
    } else if (state == ProxyRunState.ERROR) {
        actions.add("restart")
    }
    return actions
}

private fun isMissingProject(label: String): Boolean =
    label == "not in a Tyhp project" || label == "no project"

private fun lspLabel(state: LspClientState): String = when (state) {
    LspClientState.RUNNING -> "ready"
    LspClientState.STARTING -> "starting"
    LspClientState.ERROR -> "error"
    LspClientState.STOPPED -> "stopped"
}

private fun proxyLabel(state: ProxyRunState, detail: String?): String {
    val extra = detail?.trim()?.takeIf { it.isNotEmpty() }?.let { " ($it)" } ?: ""
    return when (state) {
        ProxyRunState.RUNNING -> "listening$extra"
        ProxyRunState.STARTING -> "starting"
        ProxyRunState.STOPPING -> "stopping"
        ProxyRunState.ERROR -> "error"
        ProxyRunState.STOPPED -> "stopped"
    }
}

private fun proxyTextSuffix(state: ProxyRunState): String = when (state) {
    ProxyRunState.RUNNING -> " · proxy"
    ProxyRunState.STARTING, ProxyRunState.STOPPING -> " · proxy…"
    else -> ""
}

File: tyhp-lang/vscode/src/status/statusBarModel.ts
export type StatusLspState = "stopped" | "starting" | "running" | "error";
export type StatusBinaryStatus = "ok" | "missing" | "invalid";
export type StatusProxyState = "stopped" | "starting" | "running" | "stopping" | "error";

export interface StatusBarView {
    text: string;
    tooltip: string;
    error: boolean;
    warning: boolean;
}

export interface StatusBarInput {
    projectLabel: string;
    lspState: StatusLspState;
    binaryStatus: StatusBinaryStatus;
    binaryPath?: string;
    binaryMessage?: string;
    proxyState?: StatusProxyState;
    proxyDetail?: string;
}

/**
 * Compact status-bar copy: `Tyhp` + owner project / “not in a Tyhp project” + LSP + binary health
 * + optional XDebug proxy listening state.
 */
export function formatStatusBar(input: StatusBarInput): StatusBarView {
    const project = input.projectLabel.trim() !== "" ? input.projectLabel.trim() : "not in a Tyhp project";
    const lsp = lspLabel(input.lspState);
    const binaryOk = input.binaryStatus === "ok";
    const proxyState = input.proxyState ?? "stopped";
    const proxy = proxyLabel(proxyState, input.proxyDetail);
    const missingProject = isMissingProject(project);

    const tooltipLines = [
        `Project: ${project}`,
        `Language server: ${lsp}`,
        `CLI: ${binaryOk ? input.binaryPath ?? "ok" : input.binaryMessage ?? "missing"}`,
        `XDebug proxy: ${proxy}`,
        "Click for Tyhp actions.",
    ];

    if (!binaryOk) {
        return {
            text: `$(error) Tyhp · ${project} · CLI missing`,
            tooltip: tooltipLines.join("\n"),
            error: true,
            warning: false,
        };
    }

    if (input.lspState === "error") {
        return {
            text: `$(error) Tyhp · ${project} · LSP error`,
            tooltip: tooltipLines.join("\n"),
            error: true,
            warning: false,
        };
    }

    if (input.lspState === "starting") {
        return {
            text: `$(sync~spin) Tyhp · ${project} · starting`,
            tooltip: tooltipLines.join("\n"),
            error: false,
            warning: false,
        };
    }

    if (input.lspState === "stopped") {
        return {
            text: `$(debug-disconnect) Tyhp · ${project} · LSP stopped`,
            tooltip: tooltipLines.join("\n"),
            error: false,
            warning: missingProject,
        };
    }

    if (proxyState === "error") {
        return {
            text: `$(warning) Tyhp · ${project} · proxy error`,
            tooltip: tooltipLines.join("\n"),
            error: false,
            warning: true,
        };
    }

    const proxySuffix = proxyTextSuffix(proxyState);
    const icon = missingProject ? "$(warning)" : "$(check)";
    return {
        text: `${icon} Tyhp · ${project} · ${lsp}${proxySuffix}`,
        tooltip: tooltipLines.join("\n"),
        error: false,
        warning: missingProject,
    };
}

function isMissingProject(label: string): boolean {
    return label === "not in a Tyhp project" || label === "no project";
}

function lspLabel(state: StatusLspState): string {
    switch (state) {
        case "running":
            return "ready";
        case "starting":
            return "starting";
        case "error":
            return "error";
        default:
            return "stopped";
    }
}

function proxyLabel(state: StatusProxyState, detail?: string): string {
    const extra = detail && detail.trim() !== "" ? ` (${detail.trim()})` : "";
    switch (state) {
        case "running":
            return `listening${extra}`;
        case "starting":
            return "starting";
        case "stopping":
            return "stopping";
        case "error":
            return "error";
        default:
            return "stopped";
    }
}

function proxyTextSuffix(state: StatusProxyState): string {
    switch (state) {
        case "running":
            return " · proxy";
        case "starting":
        case "stopping":
            return " · proxy…";
        default:
            return "";
    }
}

/** Status-bar quick-pick actions for the current proxy lifecycle state. */
export function proxyStatusActions(state: StatusProxyState): Array<"start" | "stop" | "restart"> {
    const actions: Array<"start" | "stop" | "restart"> = [];
    if (state !== "running" && state !== "starting") {
        actions.push("start");
    }
    if (state === "running" || state === "starting" || state === "stopping") {
        actions.push("stop", "restart");
    } else if (state === "error") {
        actions.push("restart");
    }
    return actions;
}
